File: aoc2024/aoc20.hs
import Control.Arrow (Arrow (second))
import Control.Monad (guard)
import Data.Map (Map, (!))
import Data.Map qualified as M
import Data.Set qualified as S
import Dijkstra qualified
import System.Environment (getArgs)

add :: (Num a, Num b) => (a, b) -> (a, b) -> (a, b)
add (!x1, !y1) (!x2, !y2) = (x1 + x2, y1 + y2)

getNeighbors :: Map (Int, Int) Char -> Int -> Int -> (Int, Int) -> [((Int, Int), Int)]
getNeighbors mp minCost maxCost spot = do
  xoff <- [-maxCost .. maxCost]
  yoff <- [-maxCost .. maxCost]
  let cost = abs xoff + abs yoff
  guard $ cost >= minCost
  guard $ cost <= maxCost
  let nspot = spot `add` (xoff, yoff)
  guard $ '#' /= M.findWithDefault '#' nspot mp
  pure (nspot, cost)

minCostMap :: forall state. (Ord state) => (state -> [(state, Int)]) -> state -> Map state Int
minCostMap txn start = go (M.singleton start 0) (txn start)
  where
    go :: Map state Int -> [(state, Int)] -> Map state Int
    go mp [] = mp
    go mp xs =
      let changes = filter (\(s, c) -> maybe True (c <) (M.lookup s mp)) xs
          getNext (s, c) = second (+ c) <$> txn s
          nexts = concatMap getNext changes
       in go (M.fromList changes `M.union` mp) nexts

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "aoc20.in" else head args
  grid <- lines <$> readFile filename
  let gridMap = M.fromList $ do
        (xidx, row) <- zip [0 ..] grid
        (yidx, ch) <- zip [0 ..] row
        pure ((xidx, yidx), ch)
  let boundNbrs = getNeighbors gridMap 1 1
  let startSpot = fst $ head $ filter ((== 'S') . snd) $ M.toList gridMap
  let accept = (== 'E') . (gridMap !)
  (baseDist, endSpot) <- case Dijkstra.dijkstra boundNbrs startSpot accept of
    Nothing -> ioError (userError "No path!")
    Just (e, t) -> pure (t, e)

  let distFromStart = minCostMap boundNbrs startSpot
      distFromEnd = minCostMap boundNbrs endSpot
  let findCheatNbrs = getNeighbors gridMap 2 2
  let distCheating startd (dest, destd) = startd + destd + distFromEnd ! dest
  putStr "Part 1: "
  print $
    length $
      concatMap
        ( filter (<= baseDist - 100)
            . ( \(k, v) ->
                  map (v `distCheating`) (findCheatNbrs k)
              )
        )
        (M.toList distFromStart)
  let findCheatNbrs' = getNeighbors gridMap 2 20
  putStr "Part 2: "
  print $
    length $
      concatMap
        ( filter (<= baseDist - 100)
            . ( \(k, v) ->
                  map (v `distCheating`) (findCheatNbrs' k)
              )
        )
        (M.toList distFromStart)

File: aoc2024/aoc20.py
from aoc_util import get_data_lines

data = get_data_lines(20)

startloc = None
endloc = None
mymap = {}
for xidx, row in enumerate(data):
    for yidx, ch in enumerate(row):
        mymap[(xidx, yidx)] = ch
        if ch == "S":
            startloc = (xidx, yidx)
        if ch == "E":
            endloc = (xidx, yidx)


def get_neighbors1(spot):
    (myx, myy) = spot
    nbs = [(myx - 1, myy), (myx, myy - 1), (myx, myy + 1), (myx + 1, myy)]
    return [s for s in nbs if mymap.get(s, "$") in ".SE"]


dist_to_end_map = {}
workq = [(endloc, 0)]
while workq:
    w = workq.pop(0)
    if w[0] in dist_to_end_map:
        continue
    dist_to_end_map[w[0]] = w[1]
    for n in get_neighbors1(w[0]):
        workq.append((n, w[1] + 1))

dist_from_start_map = {}
workq = [(startloc, 0)]
while workq:
    w = workq.pop(0)
    if w[0] in dist_from_start_map:
        continue
    dist_from_start_map[w[0]] = w[1]
    for n in get_neighbors1(w[0]):
        workq.append((n, w[1] + 1))

base_dist = dist_to_end_map[startloc]


def get_neighbors2(spot):
    (myx, myy) = spot
    nbs = [(myx - 2, myy), (myx, myy - 2), (myx, myy + 2), (myx + 2, myy)]
    return [s for s in nbs if mymap.get(s, "$") in ".SE"]


goodcheats = set()
for h in dist_from_start_map:
    for n in get_neighbors2(h):
        new_dist = dist_from_start_map[h] + 2 + dist_to_end_map[n]
        if base_dist - new_dist >= 100:
            goodcheats.add((h, n))

print("Part 1:", len(goodcheats))


def get_neighbors20(spot):
    (myx, myy) = spot
    nbs = []
    for xoff in range(-20, 21):
        for yoff in range(-20, 21):
            cheatlen = abs(xoff) + abs(yoff)
            if cheatlen <= 20:
                nbs.append(((myx + xoff, myy + yoff), cheatlen))
    return [s for s in nbs if mymap.get(s[0], "$") in ".SE"]


goodcheats = set()
for h in dist_from_start_map:
    for fn in get_neighbors20(h):
        (n, nd) = fn
        new_dist = dist_from_start_map[h] + nd + dist_to_end_map[n]
        if base_dist - new_dist >= 100:
            goodcheats.add((h, n))

print("Part 2:", len(goodcheats))
